Compile MFEM externally

Sbozzolo opened the pull request:
This commit puts MFEM on the same footing as the other dependencies. Instead of being unconditionally built, it is now build when building external dependencies. This allowed me to add MFEM as a dependency managed spack.

In doing this, I found a new bug in spack:
https://github.com/spack/spack/issues/51505

So, I changed our approach to testing the spack package. Instead of adding a local repo, I overwrite palace in the `builtin` repo.

As a side effect, this makes the spack CI action even faster, because now MFEM can be cached.

Among the challenges I faced in implementing this is that hypre and mfem are currently being built with autotools (in spack). This makes certain things hard/annoying. For example, passing through the `MFEM_DATA_PATH` turned out to be a pain, so I decided to vend the two meshes instead (we are already doing that for other meshes).

It should be noted that hypre have switched to cmake with release 3.0, and that there's a PR open in spack-packages to use cmake in MFEM.

Sbozzolo commented on test/unit/test-geodata.cpp:
```diff
@@ -327,9 +326,8 @@ TEST_CASE("Default IOData", "[iodata][Serial]")
-  auto ref_tet_path = fs::path(MFEM_DATA_PATH) / "ref-tetrahedron.mesh";
-  mfem::Mesh single_tet(ref_tet_path.string());
-  mfem::ParMesh pmesh(Mpi::World(), single_tet);
+  mfem::Mesh mesh = mfem::Mesh::MakeCartesian3D(1, 1, 1, mfem::Element::TETRAHEDRON);
```
I changed the mesh here because it doesn't really matter what mesh we use, and I preferred picking something that does not depend on a file.

cameronrutherford commented on .github/workflows/spack.yml:
```diff
@@ -188,6 +189,14 @@ jobs:
+      - uses: julia-actions/setup-julia@v2
+        if: needs.filter.outputs.test == 'true'
+        with:
+          version: '1'
```
Nice!! Should we add this to the docs workflow too?

Seems like quite a bit is shared between these workflows... I wonder if there's a nice way to share these files more...

1. We could create our own action... Seems excessive
2. We could combine the workflows into one file, and run with flags...

I don't think we should really do this, but it was a thought....

Sbozzolo replied:
Yes, added to the docs as well.

cameronrutherford commented on test/unit/mesh/ref-tetrahedron.mesh:
I assume this file is fine to add and is just a canonical mesh?

cameronrutherford commented on palace/CMakeLists.txt:
```diff
@@ -112,6 +112,15 @@ list(APPEND CMAKE_MODULE_PATH "${CMAKE_SOURCE_DIR}/cmake")
+  if(NOT TARGET MPI::MPI_C AND TARGET MPI::MPI_CXX)
+    add_library(MPI::MPI_C ALIAS MPI::MPI_CXX)
+  endif()
+endif()
```
Odd that this isn't the `find_package` default behavior given the MPI C++ bindings were deprecated in 3.0 - https://www.mpi-forum.org/docs/mpi-2.2/mpi22-report/node32.htm

Sbozzolo replied:
Interesting! I am not sure about this. We use MPI_CXX in a couple of other places

cameronrutherford commented on palace/CMakeLists.txt:
```diff
@@ -141,15 +150,20 @@ if(NOT "${MFEM_DIR}" STREQUAL "")
-# if(MFEM_CXX_FLAGS)
-#   # Pull compiler flags from MFEM for OpenMP and optimizations
-#   set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} ${MFEM_CXX_FLAGS}")
-# endif()
```
Why was this there in the first place 🤔

cameronrutherford reviewed (commented):
I might have missed something... LGTM! I think largely just clarifying questions.

hughcars commented on .github/workflows/spack.yml:
```diff
@@ -51,8 +51,9 @@ jobs:
-        spack_version: [develop] # ,v0.23.1]
         mpi: [openmpi] #, mpich, mvapich]
+        eigensolver: [arpack, slepc]
+        solver: [mumps, strumpack, superlu-dist]
```
Just adding in `arpack` or `strumpack` or `mumps` into the build is not going to change from the default `superlu-dist` `slepc` combination, as they are the defaults are all in. So if the goal is testing those, the defaults need to be explicitly `~`'d.

Sbozzolo replied:
I added them mostly to check that they compile correctly, but I will try translating our linux matrix to a spack matrix (given that it has better caching -> it's faster), and maybe in the future we can reduce the size of linux matrix.

Sbozzolo commented on test/unit/CMakeLists.txt:
```diff
@@ -86,17 +86,13 @@ set_property(
-  ${CMAKE_CURRENT_SOURCE_DIR}/test-geodata.cpp
-  ${CMAKE_CURRENT_SOURCE_DIR}/test-rap.cpp
-  APPEND PROPERTY COMPILE_DEFINITIONS "MFEM_DATA_PATH=\"${CMAKE_INSTALL_PREFIX}/share/palace/test/mfem-data\""
-)
```
The reason I had to do this is that MFEM is currently built with autotools (when installed with spack). For this reason, the information on `MFEM_DATA_PATH` is lost, and it is very hard recover. 

I'd be inclined to push this to the future, possibly tackling #506 at the same time.

Sbozzolo commented on CHANGELOG.md:
```diff
@@ -68,9 +68,23 @@ The format of this changelog is based on
 
+#### Build system
+
+  - When building on a machine with a GPU, `CMAKE_CUDA_ARCHITECTURES` has now a
```
Added

Sbozzolo commented on spack_repo/local/packages/palace/package.py:
```diff
@@ -144,6 +144,33 @@ class Palace(CMakePackage, CudaPackage, ROCmPackage):
+        depends_on("mfem+gslib", when="+gslib")
+        depends_on("mfem~gslib", when="~gslib")
+
+        depends_on("mfem+exceptions", when="+tests")
```
Got it, thanks!

Sbozzolo commented on .github/workflows/docs.yml:
```diff
@@ -36,13 +36,19 @@ jobs:
+      - name: Overwrite builtin Palace package
+        run: |
+          cp -r spack_repo/local/packages/palace "$(spack location --repo builtin)/packages"
+          cp extern/patch/mfem/* "$(spack location --repo builtin)/packages/palace"
```
I added developer documentation to discuss this

cameronrutherford commented on .github/workflows/spack.yml:
```diff
@@ -48,52 +48,160 @@ jobs:
+          solver: +mumps
+          cuda: ~cuda
+
+        # Fails building superlu with pointer conversion error int vs int_t
```
Should this be a bug report?

cameronrutherford commented on .github/workflows/spack.yml:
```diff
@@ -48,52 +48,160 @@ jobs:
+        #   solver: +superlu-dist
+        #   cuda: ~cuda
+
+        # Fails with undefined reference to symbol 'sqrt@@GLIBC_2.17' in libmetis
```
Should this be a bug report? Sounds like a math library linkage issue...

Sbozzolo replied:
Same comment as above on good bug reports.

cameronrutherford commented on .github/workflows/spack.yml:
```diff
@@ -48,52 +48,160 @@ jobs:
+            EIGEN_VARIANTS="+arpack~slepc"
+          else
+            EIGEN_VARIANTS="+slepc~arpack"
+          fi
```
Would enabling both be interesting / relevant?

cameronrutherford commented on .github/workflows/spack.yml:
```diff
@@ -105,19 +213,33 @@ jobs:
+              mpi:
+                require: ${{ matrix.mpi }}
+              blas:
+                require: ${{ matrix.math-libs }} threads=openmp
```
We can enforce the `threads=openmp` at the `package.py` level if it's critical.

cameronrutherford commented on docs/src/developer/spack.md:
```diff
@@ -0,0 +1,170 @@
+to the `develop` section of your `spack.yaml` pointing to your local MFEM
+checkout. Spack will then manage both packages in development mode, rebuilding
+each as needed when you make changes.
+
```
You could also include some note here if you want about using the CMake based build after the first spack build. i.e. you can do something like this once you have the first build done and you are just modifying Palace source code (very rough commands from memory):

```bash
cd ~/repos/palace
cd palace-<spack-hash>-build
source spack-build-env.txt
# optionally
make clean
make -j
``` 

This way you can iterate without calling out to Spack / Python interpreter which is moderately faster when doing small / frequent iterations - certainly way faster now that this doesn't trigger an MFEM rebuild.

cameronrutherford commented on docs/src/developer/spack.md:
````diff
@@ -0,0 +1,170 @@
+spack repo add ~/repos/palace/spack_repo/local
+```
+
+You should see a confirmation message like `==> Added repo with namespace 'local'.` The namespace `local` is significant because it allows you to
````
NIT - your line length is too long... Maybe we need a markdown formatter.

cameronrutherford commented on docs/src/developer/spack.md:
```diff
@@ -0,0 +1,170 @@
+
+#### Working Around a Spack Bug with Patches
+
+There's currently a known issue in Spack version 1.1.0 that affects packages
```
```suggestion
There's currently a known issue in Spack version 1.0.1 that affects packages
```

I think this version is wrong... Maybe you also actually mean `1.0.2`?

cameronrutherford commented on docs/src/developer/spack.md:
```diff
@@ -0,0 +1,170 @@
+
+With your local repository configured, you can now iterate on *Palace*'s
+`package.py` file. After making changes, you can test them by installing
+*Palace* again. You can also use `spack spec local.palace@develop` to see how
```
I think `spack spec` and `spack concretize` would be doing the same thing here, but `spack spec` is nice because you get more / better information IIRC... Not sure if you should change anything here...

Sbozzolo commented on .github/workflows/spack.yml:
```diff
@@ -48,52 +48,160 @@ jobs:
-        blas: [openblas] # ,intel-oneapi-mkl]
-        spack_version: [develop] # ,v0.23.1]
-        mpi: [openmpi] #, mpich, mvapich]
+        include:  # Pairwise testing (from generate_test_matrix.py)
```
`generate_test_matrix.py` uses `allpairspy` to combine options to generate good coverage with reduced number of tests. We used `allpairspy` for our `linux` tests as well, but the script was not saved, making the matrix not reproducible and hiding the underlying constraints. 

I added `generate_test_matrix` so that it is clear how the test matrix was chosen and what assumptions went in. This becomes relevant if one wants to add/remove options down the line.

Sbozzolo commented on .github/workflows/spack.yml:
```diff
@@ -48,52 +48,160 @@ jobs:
+          if [[ "${{ matrix.solver }}" == "+strumpack" ]]; then
+            SOLVER_VARIANTS="+strumpack~superlu-dist~mumps"
+          elif [[ "${{ matrix.solver }}" == "+mumps" ]]; then
+            SOLVER_VARIANTS="+mumps~superlu-dist~strumpack"
```
No, this is Hugh's comment from above: if we don't disable certain solvers, the tests won't use the other ones (because we have certain defaults and we don't change the JSON files). We must disable certain solver so that we can test the others.

cameronrutherford replied:
I agree in terms of a runtime test, but sometimes there are weird bugs that can happen at just the build stage, not the test stage. It would be unfortunate if someone added a bug that breaks the build with both solvers enabled, forcing you to have 2 builds to benchmark and compare solvers.

It would be an even more obscure bug to happen at the test stage as well.

Sbozzolo commented on spack_repo/local/packages/palace/package.py:
```diff
@@ -50,17 +50,23 @@ class Palace(CMakePackage, CudaPackage, ROCmPackage):
+            when="@0.15:")
+
+    # TODO: Add lcov and other packages that might be needed to truly measure
+    # coverage.
```
This package.py turns on the one cmake knob needed to enable coverage in Palace. However, there might be extra packages needed to truly extract coverage and make and html report. This is partially related to #538, so we will need to see.

Sbozzolo commented on spack_repo/local/packages/palace/package.py:
```diff
@@ -50,17 +50,23 @@ class Palace(CMakePackage, CudaPackage, ROCmPackage):
     depends_on("c", type="build")
     depends_on("cxx", type="build")
-    depends_on("cmake@3.21:", type="build")
+    depends_on("cmake@3.24:", type="build")
```
This was missed in #398

cameronrutherford commented on .github/workflows/spack.yml:
```diff
@@ -48,52 +48,160 @@ jobs:
+          cuda: ~cuda
 
-    runs-on: palace_ubuntu-latest_16-core
+        # Fails configuring PETSc
```
I don't think it's unreasonable to at least open a bug report with a link to the failing pipeline, even if you don't have a strong reproducer / minimum example. The more complex example that runs in a GitHub workflow is still useful as a bug report to our dependent projects IMO, and it's easy enough for them to reproduce if they spin up their own GitHub action.

I am also suggesting this as it would be good to have some tracking of the bug... If we don't open a downstream issue, then we should open an issue in Palace to track the failing builds?

That, or we should minimize and officially state our supported platforms.

Sbozzolo commented on docs/src/developer/spack.md:
```diff
@@ -0,0 +1,170 @@
+packages like compilers, MPI implementations, and common libraries. Spack will
+then use these instead of building them from scratch, which can save hours on
+the first build.
+
```
Yes, I think this is for the future

cameronrutherford commented on .github/workflows/spack.yml:
```diff
@@ -105,19 +213,33 @@ jobs:
+              # -fPIC seems to be required because metis always compiles some shared objects
+              # -lc seems to be required for intel compilers
+              metis:
+                require: cflags=-fPIC cppflags=-fPIC ldflags=-lc
```
Hmmm I think that would suggest that this is an edge case bug with the intel compilers, but to your point, hard to narrow down past that. It would be nice to verify this and perhaps only conditionally apply these things based on compiler, but that's probably overkill.

Sbozzolo commented on spack_repo/local/packages/palace/package.py:
```diff
@@ -156,14 +197,16 @@ class Palace(CMakePackage, CudaPackage, ROCmPackage):
         depends_on("sundials~cuda", when="~cuda")
         depends_on("sundials~rocm", when="~rocm")
 
+    conflicts("+tests", when="@:0.14", msg="Building tests is only supported for Palace versions after 0.14")
```
Yes, I think it's good to have this error message

Sbozzolo commented:
I found that some static builds were broken (because our build system relied on transitively linking libraries through MFEM) and fixed that.

In this, I also found that the package.py for SUNDIALS in the released `spack-packages` is broken because it refers to an non-existing commit. This was fixed [here](https://github.com/spack/spack-packages/commit/9b37b243d47faed5d54dd194f6e7d15332cc8fcb), but it is not part of the released packages. So, as of right now, I have resorted to just copying the spack recipes we need with the fixes we need. We can discuss how to best do this.

Following @cameronrutherford's suggestion, I added an `env` file and managed to get another combination to build, but one of the regression tests fails: https://github.com/awslabs/palace/actions/runs/19281378654/job/55133032500
I tried reproducing the same palace spec locally to see if the issue was with the choice of solvers, but the test passes locally.

I also think I might have found an issue with the PETSc configure error in `-lomp` not being passed depending on the config for petsc/superlu, but I could not verify this on GitHub actions.

hughcars commented on docs/src/developer/spack.md:
```diff
@@ -0,0 +1,173 @@
+
+If you see MFEM's package information, you're all set. This workaround ensures
+that when Spack processes patches for *Palace*, it can find MFEM in the same
+repository namespace and avoid the bug.
```
Didn't get added on the review, but the spirit is in the main message:

Tried this out to do an update of the mfem dependency, and one quirk is doing `spack develop mfem` starts the aim from master, which causes the patching to fail. It's easy enough going into the folder and doing `git checkout <sha>` if you know where the sha is kept `ExternalGitTags` or the spack recipe, but it doesn't work out the box. 

Given this is one of the real strengths of this external mfem development, it might be worth adding a small section here mentioning this small pitfall, and maybe also that `spack concretize --force` is especially necessary if changing the patches themselves

Sbozzolo commented on docs/src/developer/spack.md:
````diff
@@ -0,0 +1,173 @@
+spack repo add ~/repos/palace/spack_repo/local
+```
+
+You should see a confirmation message like `==> Added repo with namespace 'local'.` The namespace `local` is significant because it allows you to
````
I think this is something that is handled by out `format-source` script. In some cases, it forces some long lines.

Sbozzolo commented on docs/src/developer/spack.md:
```diff
@@ -0,0 +1,173 @@
+repository as well.
+
+Rather than maintaining a duplicate copy of the MFEM recipe, you can create a
+symbolic link from Spack's built-in MFEM package to your local repository:
```
I removed the local copy in our spack repo and do everything with links now.

Sbozzolo commented on spack_repo/local/packages/sundials/package.py:
```diff
@@ -0,0 +1,798 @@
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+import os
+import sys
```
Removed all of these in favor of PRs and patches.

Sbozzolo commented:
I think I addressed the comments and changed to base to be #578 (which, in turn, is against #498). This also verified that the change compiles fine on GPUs and with the new hypre.

Some Cmake work was needed to make sure everything compiled fine with static builds as well (before, MFEM would carry a bunch of dependencies transitively).

cameronrutherford commented on CHANGELOG.md:
```diff
@@ -87,6 +87,20 @@ The format of this changelog is based on
+    530](https://github.com/awslabs/palace/pull/530).
+  - The spack Palace build now supports a new variant `+tests` to compile and
+    install the unit tests. [PR
+    549](https://github.com/awslabs/palace/pull/549).
```
I think this isn't true anymore as we now support the `--test root` style instead of it being an explicit variant.

cameronrutherford commented on .github/actions/setup-runner/action.yml:
```diff
@@ -101,6 +101,26 @@ runs:
+      run: |
+        PALACE_DIR=$PWD
+        cd $(spack location --repo builtin)
+        for patch in "$PALACE_DIR"/spack_repo/patches/*.diff; do
```
Is this directory correct?
```suggestion
        for patch in "$PALACE_DIR"/spack_repo/local/packages/palace/*.diff; do
```

cameronrutherford commented on .github/workflows/docs.yml:
```diff
@@ -36,18 +36,33 @@ jobs:
           buildcache: true
           color: true
 
+      - name: Patch Spack Packages for known problems
```
Maybe this should explicitly say that this is MFEM related.

cameronrutherford commented on .github/actions/setup-runner/action.yml:
```diff
@@ -101,6 +101,26 @@ runs:
+        done
+
+        # Workaround for https://github.com/spack/spack/issues/51505
+        ln -s "$(spack location --repo builtin)"/packages/mfem "$PALACE_DIR"/spack_repo/local/packages/
```
I think the symlink is necessary, but I would hope that spack is the thing responsible for applying said patches, not the action itself.

cameronrutherford commented on palace/CMakeLists.txt:
```diff
@@ -52,9 +57,36 @@ set(SLEPC_DIR "" CACHE STRING
-if(PALACE_WITH_ARPACK)
+if(PALACE_WITH_ARPACK OR PALACE_WITH_STRUMPACK OR PALACE_WITH_MUMPS)
   enable_language(Fortran)
 endif()
```
I think this might need to also be updated in our `package.py`? IIRC we only have `+fortran` enabled for one solver... Might be wrong 😅

cameronrutherford commented on palace/CMakeLists.txt:
```diff
@@ -330,14 +384,217 @@ if(PALACE_WITH_OPENMP)
+endif()
+
+# Link ZLIB
+find_package(ZLIB)
```
Or perhaps a dependency provides it in their exported CMake...

cameronrutherford reviewed (changes requested):
Looking good!! Couple of requests but I think this is in great shape.

P.S. I accidentally closed the tab while writing the review and so some comments might've been duplicated / lost.

Sbozzolo commented on spack_repo/local/packages/palace/package.py:
```diff
@@ -235,6 +266,18 @@ def cmake_args(self):
+            args.append(self.define("METIS_DIR", self.spec["metis"].prefix))
+            args.append(self.define("PARMETIS_DIR", self.spec["parmetis"].prefix))
+            args.append(self.define("HYPRE_DIR", self.spec["hypre"].prefix))
+
```
Yes, that's exactly what I am doing.
